changefeedccl: allow users to perform initial scans on newly added targets

sherman-grewal opened the pull request:
changefeedccl: allow users to perform initial scans on
newly added targets in the ALTER CHANGEFEED statement

References #75895

Currently, when a new target is added to an existing
changefeed through the ALTER CHANGEFEED statement,
no initial scan is performed on the newly added
target. With this change, users can indicate that
they want an initial scan by executing the following
statement:

ALTER CHANGEFEED <job_id> ADD <targets> WITH initial_scan

Users can also explicitly ask for no initial scan by
replacing 'initial_scan' with 'no_initial_scan'.
The default behavior is to perform no initial scans
when new targets are added.

Release note (enterprise change): Users can now
perform initial scans on newly added targets
by executing the following statement:

ALTER CHANGEFEED <job_id> ADD <targets> WITH initial_scan

The default behavior is to perform no initial scans
on newly added targets, but users can explicitly
request this by replacing the 'initial_scan' with
'no_initial_scan'.

Release justification: low danger enhancement to newly
implemented functionality.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/77263)
<!-- Reviewable:end -->

miretskiy reviewed (changes requested):
> Reviewed 6 of 7 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/77263)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @samiskin and @sherman-grewal)

___
*[-- commits, line 3 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/77263#-MxA6a-f6VcO139OaNJt):*
You need to expand this message;

We now also require "Release Justification" to explain why you want to merge (since we're in a stability period).
You can use something like: "Release Justification: low danger enhancement to newly implemented functionality".
___
*[pkg/ccl/changefeedccl/alter_changefeed_stmt.go, line 319 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/77263#-MxA9get2vxhKQfTNUKO:-MxA9get2vxhKQfTNUKP:b-fse8jx) ([raw file](https://github.com/cockroachdb/cockroach/blob/5d5b2595d29b36442346084f75135d8f3cb133f2/pkg/ccl/changefeedccl/alter_changefeed_stmt.go#L319)):*
> ```Go
> 
> 		progress := jobspb.Progress{
> 			Progress: &jobspb.Progress_HighWater{},
> ```

I'm think this is not right; Let's discuss in person.


<!-- Sent from Reviewable.io -->

miretskiy reviewed (changes requested):
This is getting so close.  I still have few comments, mostly nits.  This looks great!

> Reviewed 1 of 7 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/77263#-:-MxtTsFI7Aslv9Ssgb9L:bv5gsei)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @HonoreDB, @miretskiy, @samiskin, and @sherman-grewal)

___
*[pkg/ccl/changefeedccl/alter_changefeed_stmt.go, line 92 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/77263#-MxtM-eV97u4hAvFXM43:-MxtM-eV97u4hAvFXM44:b-u37nfi) ([raw file](https://github.com/cockroachdb/cockroach/blob/a75223fabbb6a29f0288fdcdf71cc6d8597e82ca/pkg/ccl/changefeedccl/alter_changefeed_stmt.go#L92)):*
> ```Go
> 		}
> 
> 		sinkURIFn, err := p.TypeAsString(ctx, newChangefeedStmt.SinkURI, `ALTER CHANGEFEED`)
> ```

I'm confused: why do we need to evaluate sinkURI again?  if newSinkURI was returned by generateNewOpts -- it must be a DString - i.e. fully evaluated string.  If you got nil for newSinkURI -- that's fine, just set it to prevDetails.SinkURI as you do now.
___
*[pkg/ccl/changefeedccl/alter_changefeed_stmt.go, line 97 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/77263#-MxtMoLr-LtOXc7Ybzg5:-MxtMoLr-LtOXc7Ybzg6:bd94dqt) ([raw file](https://github.com/cockroachdb/cockroach/blob/a75223fabbb6a29f0288fdcdf71cc6d8597e82ca/pkg/ccl/changefeedccl/alter_changefeed_stmt.go#L97)):*
> ```Go
> 		}
> 
> 		optsFn, err := p.TypeAsStringOpts(ctx, newChangefeedStmt.Options, changefeedbase.ChangefeedOptionExpectValues)
> ```

again, not sure why we need another opts eval.  Please check if we need this.
___
*[pkg/ccl/changefeedccl/alter_changefeed_stmt.go, line 112 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/77263#-MxtN25T00tBsWYWzT7r:-MxtN25T00tBsWYWzT7s:bopmson) ([raw file](https://github.com/cockroachdb/cockroach/blob/a75223fabbb6a29f0288fdcdf71cc6d8597e82ca/pkg/ccl/changefeedccl/alter_changefeed_stmt.go#L112)):*
> ```Go
> 		}
> 
> 		newTargets, newProgress, newStatementTime, err := generateNewTargets(ctx, p, alterChangefeedStmt.Cmds, opts, job)
> ```

nit: wrap if you can?
___
*[pkg/ccl/changefeedccl/alter_changefeed_stmt.go, line 303 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/77263#-MxtOG5ZC0qwNYUqcpnB:-MxtOG5ZC0qwNYUqcpnC:bbvvl73) ([raw file](https://github.com/cockroachdb/cockroach/blob/a75223fabbb6a29f0288fdcdf71cc6d8597e82ca/pkg/ccl/changefeedccl/alter_changefeed_stmt.go#L303)):*
> ```Go
> 	opts map[string]string,
> 	job *jobs.Job,
> ) (*tree.TargetList, *jobspb.Progress, *hlc.Timestamp, error) {
> ```

probably makes sense to return hlc.Timestamp -- simple structs should be returned as structs, not as pointers.
Besides, you assume at the call site, that (if err != nil) this time is always set 
___
*[pkg/ccl/changefeedccl/alter_changefeed_stmt.go, line 304 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/77263#-MxtNUPB--zY4Is-RQrA:-MxtNUPB--zY4Is-RQrB:b-pzfgz9) ([raw file](https://github.com/cockroachdb/cockroach/blob/a75223fabbb6a29f0288fdcdf71cc6d8597e82ca/pkg/ccl/changefeedccl/alter_changefeed_stmt.go#L304)):*
> ```Go
> 	job *jobs.Job,
> ) (*tree.TargetList, *jobspb.Progress, *hlc.Timestamp, error) {
> 	prevDetails := job.Details().(jobspb.ChangefeedDetails)
> ```

I would prefer to pass in prevDetails/prevProgress to this method instead of passing job.
(seems like a nit: but really, job is a wrong level of abstraction here -- you get old details/progress and return
new set).
___
*[pkg/ccl/changefeedccl/alter_changefeed_stmt.go, line 319 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/77263#-MxtOcwMDBYZguA3EK7W:-MxtOcwMDBYZguA3EK7X:bj8i9dl) ([raw file](https://github.com/cockroachdb/cockroach/blob/a75223fabbb6a29f0288fdcdf71cc6d8597e82ca/pkg/ccl/changefeedccl/alter_changefeed_stmt.go#L319)):*
> ```Go
> 	}
> 
> 	allDescs, err := backupresolver.LoadAllDescs(ctx, p.ExecCfg(), statementTime)
> ```

I think this is incorrect.  We are using the current time to load all of the descriptors.  What happens if we add new table that was created after changefeed was paused? Wouldn't we fail the changefeed because it could resume from the prevDetails.StatementTime and try to resolve this new descriptor  as of that time -- and that will fail?

Unless... i'm wrong.  And validate targets does that for us...  If that's the case, then comment why we resolve as of current time.
___
*[pkg/ccl/changefeedccl/alter_changefeed_stmt.go, line 387 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/77263#-MxtPRsv55wSAznu0F4q:-MxtPRsv55wSAznu0F4r:bxs5bac) ([raw file](https://github.com/cockroachdb/cockroach/blob/a75223fabbb6a29f0288fdcdf71cc6d8597e82ca/pkg/ccl/changefeedccl/alter_changefeed_stmt.go#L387)):*
> ```Go
> 			// targets. Hence, the user must explicitly state that they want an
> 			// initial scan performed on the new targets
> 			newJobProgress, newJobStatementTime, err = generateNewProgress(newJobProgress, newJobStatementTime, existingTargetSpans, addedTargetSpans, withInitialScan)
> ```

nit: wrap
___
*[pkg/ccl/changefeedccl/alter_changefeed_stmt.go, line 408 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/77263#-MxtTJj5EwPgNxLmM54Z:-MxtTJj5EwPgNxLmM54_:bsi9qdc) ([raw file](https://github.com/cockroachdb/cockroach/blob/a75223fabbb6a29f0288fdcdf71cc6d8597e82ca/pkg/ccl/changefeedccl/alter_changefeed_stmt.go#L408)):*
> ```Go
> 				return nil, nil, nil, err
> 			}
> 			removeSpansFromProgress(newJobProgress, droppedTargetSpans)
> ```

does it make sense to collect all targets to add, and all targets to remove and call add/remove outside this loop?
Or it doesn't matter -- logic works anyway, and probably not worth optimizing because we will add few (usually 1) target at a time anyway?
___
*[pkg/ccl/changefeedccl/alter_changefeed_stmt.go, line 456 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/77263#-MxtT8Mk3MPtGYukSiwl:-MxtT8Mk3MPtGYukSiwm:b-ckf8wb) ([raw file](https://github.com/cockroachdb/cockroach/blob/a75223fabbb6a29f0288fdcdf71cc6d8597e82ca/pkg/ccl/changefeedccl/alter_changefeed_stmt.go#L456)):*
> ```Go
> 	}
> 
> 	for _, targetName := range newTargets.Tables {
> ```

where do we validate that there is at least 1 target remaining?
___
*[pkg/ccl/changefeedccl/alter_changefeed_stmt.go, line 465 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/77263#-MxtSPnWA90dOrfJf6m7:-MxtSPnWA90dOrfJf6m8:b-mxyd3x) ([raw file](https://github.com/cockroachdb/cockroach/blob/a75223fabbb6a29f0288fdcdf71cc6d8597e82ca/pkg/ccl/changefeedccl/alter_changefeed_stmt.go#L465)):*
> ```Go
> 				return errors.Errorf(`target %q cannot be resolved as of the high water mark. Please wait until the high water mark progresses past the creation time of this target in order to add it to the changefeed.`, tree.ErrString(targetName))
> 			}
> 			return errors.Errorf(`target %q cannot be resolved as of the creation time of the changefeed. Please wait until the high water mark progresses past the creation time of this target in order to add it to the changefeed.`, tree.ErrString(targetName))
> ```

can we wrap these error message?
you could do
```
  return errors.Newf("blahh...." +
      " more blah " +
      " ....."
```
___
*[pkg/ccl/changefeedccl/alter_changefeed_stmt.go, line 477 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/77263#-MxtSAAWDdFyD59S4qod:-MxtSAAX5QABuI6A59tF:bnjwu18) ([raw file](https://github.com/cockroachdb/cockroach/blob/a75223fabbb6a29f0288fdcdf71cc6d8597e82ca/pkg/ccl/changefeedccl/alter_changefeed_stmt.go#L477)):*
> ```Go
> // no_initial_scan), and the current status of the job. If the progress does not
> // need to be updated, we will simply return the previous progress and statement
> // time that is passed into the function
> ```

nit: '.' at the end of sentence... Here and in other comments in this function.
___
*[pkg/ccl/changefeedccl/alter_changefeed_stmt.go, line 505 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/77263#-MxtRU9U9hpgtwWABfOH:-MxtRU9U9hpgtwWABfOI:b-4pyr) ([raw file](https://github.com/cockroachdb/cockroach/blob/a75223fabbb6a29f0288fdcdf71cc6d8597e82ca/pkg/ccl/changefeedccl/alter_changefeed_stmt.go#L505)):*
> ```Go
> 		// if we do not want to perform an initial scan on these targets, update the
> 		// checkpoint to include the spans of these targets so that the changefeed
> 		// will skip performing a backfill on these targets
> ```

Very nice comment.
___
*[pkg/ccl/changefeedccl/alter_changefeed_stmt.go, line 520 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/77263#-MxtR0sB0ypQy-izmUte:-MxtR0sB0ypQy-izmUtf:b-2u3880) ([raw file](https://github.com/cockroachdb/cockroach/blob/a75223fabbb6a29f0288fdcdf71cc6d8597e82ca/pkg/ccl/changefeedccl/alter_changefeed_stmt.go#L520)):*
> ```Go
> 
> 	// if the checkpoint is non-empty
> 	if changefeedProgress != nil && changefeedProgress.Checkpoint != nil && len(changefeedProgress.Checkpoint.Spans) != 0 {
> ```

readability nit: Lets introduce on top of this function 2 variables, and use those here and above:
```
var haveHighwater := !(prevHighWater == nil || prevHighWater.IsEmpty())
var haveCheckpoint := changefeedProgress != nil && changefeedProgress.Checkpoint != nil && len(changefeedProgress.Checkpoint.Spans) != 0 
```
___
*[pkg/ccl/changefeedccl/alter_changefeed_stmt.go, line 523 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/77263#-MxtRaE36FscWLhgmhHO:-MxtRaE36FscWLhgmhHP:btw7l7o) ([raw file](https://github.com/cockroachdb/cockroach/blob/a75223fabbb6a29f0288fdcdf71cc6d8597e82ca/pkg/ccl/changefeedccl/alter_changefeed_stmt.go#L523)):*
> ```Go
> 		// if we want to perform an initial scan on the new targets
> 		if withInitialScan {
> 			return prevProgress, prevStatementTime, errors.Errorf(`cannot perform initial scan on newly added targets during a non-initial backfill, please wait for the backfill to complete before adding these targets.`)
> ```

nit: mega wrap wanted.

nit2: you probably want to say that they need to unpause this changefeed, and wait... until something... Probably for high water mark to be larger than the current value.
___
*[pkg/ccl/changefeedccl/alter_changefeed_stmt.go, line 555 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/77263#-MxtSF-04tEfW7vW9LV1:-MxtSF-04tEfW7vW9LV2:bm2awly) ([raw file](https://github.com/cockroachdb/cockroach/blob/a75223fabbb6a29f0288fdcdf71cc6d8597e82ca/pkg/ccl/changefeedccl/alter_changefeed_stmt.go#L555)):*
> ```Go
> 	// To avoid this, we update the statement time of the job to the previous high
> 	// watermark, and add all the existing targets to the checkpoint to skip the
> 	// initial scan on these targets
> ```

:heart: this comment.  
nit: "." at the end.
___
*[pkg/ccl/changefeedccl/alter_changefeed_stmt.go, line 572 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/77263#-MxtTmqs-Dz-uYBH8M_G:-MxtTmqs-Dz-uYBH8M_H:bi2st0d) ([raw file](https://github.com/cockroachdb/cockroach/blob/a75223fabbb6a29f0288fdcdf71cc6d8597e82ca/pkg/ccl/changefeedccl/alter_changefeed_stmt.go#L572)):*
> ```Go
> }
> 
> func removeSpansFromProgress(prevProgress jobspb.Progress, spans []roachpb.Span) {
> ```

nit: rename spans to spansToRemove?


<!-- Sent from Reviewable.io -->

samiskin commented on pkg/ccl/changefeedccl/alter_changefeed_stmt.go:
```diff
@@ -305,15 +167,443 @@ func alterChangefeedPlanHook(
+// no_initial_scan), and the current status of the job. If the progress does not
+// need to be updated, we will simply return the previous progress and statement
+// time that is passed into the function
+func generateNewProgress(
```
I think this function would flow better readability-wise if it was structured as:
```go
inInitialBackfill := prevHighWater == nil || prevHighWater.IsEmpty()
inSchemaChangeBackfill := !inInitialBackfill && ....

if withInitialBackfill {
  // No changes needed, the new spans will join the existing backfill
  if InBackfill {
    return prevProgress, prevStatementTime, nil
  }

  // ... comment explaining why this wouldn't work
  if inSchemaChangeBackfill {
    // return error
  }

  // If we'd like to perform an initial scan on new targets ....
  newStatementTime = *prevHighWater
  newProgress := ... 
  return newProgress, newStatementTime, nil
}


// If we don't want a scan of new targets but we're in an ongoing backfill,
// ensure new spans are skipped by adding it to the checkpoint
if inInitialBackfill || inSchemaChangeBackfill {
    var expandedCheckpoint roachpb.SpanGroup
    ... // initialize expanded checkpoint

    newProgress := ... 
    return newProgress, prevStatementTime, nil
}
  
// Without any backfills to consider, no changes are needed
return prevProgress, prevStatementTime, nil

```

Not that you'd need to do _exactly_ this, but friction I ran into reading the current flow that I wanted to reduce was:
- Reading later code, seeing `spanGroup`, going "wait what is that", then scrolling up to figure out that it meant "all the old checkpointed spans and new spans", and then having to read each block to confirm that "ok this is only used as an expanded checkpoint".  Having it be called `expandedCheckpoint` makes it easier to infer without scrolling to its definition, and having it be scoped to the context where its always used helps the reader to know that its not relevant information anywhere else.
- Reading the contents of the first if statement going "ok why does this make sense if the high water is empty -> oh ok its cause we're in a backfill", and the second going "ok this block is for an empty checkpoint, but the error says its a schema change, and ok looks like this block shoudl only run for a schema change since we've handled the initial backfill already".  The condition is using specific data to infer an environment (some backfill), while the contents of the block are operating under the assumption of that entire environment rather than just that specific data, so it helps to make the data->environment link explicit by moving them into their own variables.  It also helps highlight potential issues, such as "we're saying its a schema change backfill if theres a checkpoint outside of initial scan, but what if this runs during a schema change backfill but before a checkpoint is laid down?".  

---

And on that note, have we tested that case out?  Would we just have 2 backfills of the new targets, one as it approaches statement time and another once it sees the BACKFILL boundary event to trigger the schema backfill?  Would the checkpoint from the initial backfill on the old spans be cleared out so as to not interfere with the schema backfill?

sherman-grewal replied:
Thanks for the suggestion! I updated the function and I believe it should be easier to understand now. Also with the case that you mentioned in this comment, since we are adding targets with a non-empty high watermark and empty checkpoint, we will:
* reset the high watermark
* set the statement time to the previous high watermark
* add all existing target spans to the checkpoint

As a result, I believe that this case would be equivalent to the case in which we are performing an original initial scan and a schema change occurs during that initial scan.

samiskin commented on pkg/ccl/changefeedccl/alter_changefeed_stmt.go:
```diff
@@ -305,15 +167,443 @@ func alterChangefeedPlanHook(
+	// update the progress of the changefeed since there are no backfills in
+	// progress already
+	if !withInitialScan {
+		return prevProgress, prevStatementTime, nil
```
Does this mean that the "Statement Time" of a changefeed now means "Either when you initially ran `CREATE CHANGEFEED`, or the last time you ran `ALTER CHANGEFEED ADD ... WITH initial_scan`, but not if you ran `ALTER CHANGEFEED` without an initial scan"

sherman-grewal replied:
yeah that is right. As discussed offline, we're going to rename the StatementTime to ScanTime so that the field name stays consistent with its meaning.

sherman-grewal reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/77263)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @HonoreDB, @miretskiy, @samiskin, and @sherman-grewal)

___
*[-- commits, line 3 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/77263#-MxA6a-f6VcO139OaNJt):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

You need to expand this message;

We now also require "Release Justification" to explain why you want to merge (since we're in a stability period).
You can use something like: "Release Justification: low danger enhancement to newly implemented functionality".
</blockquote></details>

Done.
___
*[pkg/ccl/changefeedccl/alter_changefeed_stmt.go, line 92 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/77263#-MxtM-eV97u4hAvFXM43:-MxtyX895eNgri8rfq2L:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a75223fabbb6a29f0288fdcdf71cc6d8597e82ca/pkg/ccl/changefeedccl/alter_changefeed_stmt.go#L92)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

I'm confused: why do we need to evaluate sinkURI again?  if newSinkURI was returned by generateNewOpts -- it must be a DString - i.e. fully evaluated string.  If you got nil for newSinkURI -- that's fine, just set it to prevDetails.SinkURI as you do now.
</blockquote></details>

Done.
___
*[pkg/ccl/changefeedccl/alter_changefeed_stmt.go, line 97 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/77263#-MxtMoLr-LtOXc7Ybzg5:-MxtyXdB2ki2iN_Dy75B:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a75223fabbb6a29f0288fdcdf71cc6d8597e82ca/pkg/ccl/changefeedccl/alter_changefeed_stmt.go#L97)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

again, not sure why we need another opts eval.  Please check if we need this.
</blockquote></details>

Done.
___
*[pkg/ccl/changefeedccl/alter_changefeed_stmt.go, line 303 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/77263#-MxtOG5ZC0qwNYUqcpnB:-MxtolNsAcn5jquYJQN1:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a75223fabbb6a29f0288fdcdf71cc6d8597e82ca/pkg/ccl/changefeedccl/alter_changefeed_stmt.go#L303)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

probably makes sense to return hlc.Timestamp -- simple structs should be returned as structs, not as pointers.
Besides, you assume at the call site, that (if err != nil) this time is always set 
</blockquote></details>

Done.
___
*[pkg/ccl/changefeedccl/alter_changefeed_stmt.go, line 304 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/77263#-MxtNUPB--zY4Is-RQrA:-Mxtkj8TEVL5ys-HvDn3:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a75223fabbb6a29f0288fdcdf71cc6d8597e82ca/pkg/ccl/changefeedccl/alter_changefeed_stmt.go#L304)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

I would prefer to pass in prevDetails/prevProgress to this method instead of passing job.
(seems like a nit: but really, job is a wrong level of abstraction here -- you get old details/progress and return
new set).
</blockquote></details>

Done.
___
*[pkg/ccl/changefeedccl/alter_changefeed_stmt.go, line 319 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/77263#-MxtOcwMDBYZguA3EK7W:-Mxtklpl43I9qPP82Oa0:b736w9b) ([raw file](https://github.com/cockroachdb/cockroach/blob/a75223fabbb6a29f0288fdcdf71cc6d8597e82ca/pkg/ccl/changefeedccl/alter_changefeed_stmt.go#L319)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

I think this is incorrect.  We are using the current time to load all of the descriptors.  What happens if we add new table that was created after changefeed was paused? Wouldn't we fail the changefeed because it could resume from the prevDetails.StatementTime and try to resolve this new descriptor  as of that time -- and that will fail?

Unless... i'm wrong.  And validate targets does that for us...  If that's the case, then comment why we resolve as of current time.
</blockquote></details>

In validateNewTargets , we make sure that the new tables (if any were added) can be resolved as of the high water mark or statement time. See TestAlterChangefeedAddTargetErrors for how we handle the cases you mentioned. The reason why I resolve as of current time is because I wanted clear error messages between targets that don't exist at all vs. targets that do not exist as of the hwm/statementtime. I will add more comments to explain this reasoning
___
*[pkg/ccl/changefeedccl/alter_changefeed_stmt.go, line 408 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/77263#-MxtTJj5EwPgNxLmM54Z:-Mxtkph99fD1tMtf1F4_:b50v9me) ([raw file](https://github.com/cockroachdb/cockroach/blob/a75223fabbb6a29f0288fdcdf71cc6d8597e82ca/pkg/ccl/changefeedccl/alter_changefeed_stmt.go#L408)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

does it make sense to collect all targets to add, and all targets to remove and call add/remove outside this loop?
Or it doesn't matter -- logic works anyway, and probably not worth optimizing because we will add few (usually 1) target at a time anyway?
</blockquote></details>

We could do this, but I think it would worsen the readability of the code and I'm not sure if it would optimize the code significantly since we add a few targets at a time.
___
*[pkg/ccl/changefeedccl/alter_changefeed_stmt.go, line 456 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/77263#-MxtT8Mk3MPtGYukSiwl:-MxtktPY18C9aBNMKJy7:bi88k0k) ([raw file](https://github.com/cockroachdb/cockroach/blob/a75223fabbb6a29f0288fdcdf71cc6d8597e82ca/pkg/ccl/changefeedccl/alter_changefeed_stmt.go#L456)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

where do we validate that there is at least 1 target remaining?
</blockquote></details>

See `validateNewTargets`
```go
	if len(newTargets.Tables) == 0 {
		return pgerror.New(pgcode.InvalidParameterValue, "cannot drop all targets")
	} 
```
___
*[pkg/ccl/changefeedccl/alter_changefeed_stmt.go, line 465 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/77263#-MxtSPnWA90dOrfJf6m7:-Mxtq8Km0gG-WjISIUuT:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a75223fabbb6a29f0288fdcdf71cc6d8597e82ca/pkg/ccl/changefeedccl/alter_changefeed_stmt.go#L465)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

can we wrap these error message?
you could do
```
  return errors.Newf("blahh...." +
      " more blah " +
      " ....."
```
</blockquote></details>

Done.
___
*[pkg/ccl/changefeedccl/alter_changefeed_stmt.go, line 520 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/77263#-MxtR0sB0ypQy-izmUte:-MxtlwFCApfuLQV9aj-T:b-3p6zrj) ([raw file](https://github.com/cockroachdb/cockroach/blob/a75223fabbb6a29f0288fdcdf71cc6d8597e82ca/pkg/ccl/changefeedccl/alter_changefeed_stmt.go#L520)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

readability nit: Lets introduce on top of this function 2 variables, and use those here and above:
```
var haveHighwater := !(prevHighWater == nil || prevHighWater.IsEmpty())
var haveCheckpoint := changefeedProgress != nil && changefeedProgress.Checkpoint != nil && len(changefeedProgress.Checkpoint.Spans) != 0 
```

</blockquote></details>

Done. Great suggestion!
___
*[pkg/ccl/changefeedccl/alter_changefeed_stmt.go, line 547 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/77263#-Mx87xP-rVbGb5UXS0VW-r5-547:-MxtxuzN4wXwVwG6VIDd:bw6v9nb) ([raw file](https://github.com/cockroachdb/cockroach/blob/a75223fabbb6a29f0288fdcdf71cc6d8597e82ca/pkg/ccl/changefeedccl/alter_changefeed_stmt.go#L547)):*
<details><summary><i>Previously, samiskin (Shiranka Miskin) wrote…</i></summary><blockquote>

Does this mean that the "Statement Time" of a changefeed now means "Either when you initially ran `CREATE CHANGEFEED`, or the last time you ran `ALTER CHANGEFEED ADD ... WITH initial_scan`, but not if you ran `ALTER CHANGEFEED` without an initial scan"
</blockquote></details>

Yeah this is correct. I have added a TODO to rename this field from StatementTime to ScanTime, which would then represent the time at which the changefeed performed an initial scan. No external system relies on StatementTime at the moment (e.g. the 'created' column in SHOW JOBS does not use StatementTime), so this change is safe, but we will certainly need to rename it soon.
___
*[pkg/ccl/changefeedccl/alter_changefeed_stmt.go, line 555 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/77263#-MxtSF-04tEfW7vW9LV1:-MxtmIPp0uc_k6OZLlSf:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a75223fabbb6a29f0288fdcdf71cc6d8597e82ca/pkg/ccl/changefeedccl/alter_changefeed_stmt.go#L555)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

:heart: this comment.  
nit: "." at the end.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

miretskiy reviewed (approved):
> Reviewed 3 of 6 files at r6, 1 of 1 files at r7.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/77263)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @HonoreDB, @miretskiy, @samiskin, and @sherman-grewal)

___
*[pkg/ccl/changefeedccl/alter_changefeed_stmt.go, line 515 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/77263#-My8p6w-3J56AqKKFFDS:-My8p6w-3J56AqKKFFDT:bcnukli) ([raw file](https://github.com/cockroachdb/cockroach/blob/b902e52d31dd278732a47a92d2155d61690088a1/pkg/ccl/changefeedccl/alter_changefeed_stmt.go#L515)):*
> ```Go
> 		return prevProgress, prevStatementTime, errors.Errorf(
> 			`cannot perform initial scan on newly added targets while the checkpoint is non-empty, ` +
> 				`please unpause the changefeed and wait until the high watermark progresses past the current value to add these targets.`,
> ```

nit: you can probably include the timestamp in the error message.


<!-- Sent from Reviewable.io -->

Commit:
changefeedccl: allow users to perform initial scans on
newly added targets in the ALTER CHANGEFEED statement

Currently, when a new target is added to an existing
changefeed through the ALTER CHANGEFEED statement,
no initial scan is performed on the newly added
target. With this change, users can indicate that
they want an initial scan by executing the following
statement:

ALTER CHANGEFEED <job_id> ADD <targets> WITH initial_scan

Users can also explicitly ask for no initial scan by
replacing 'initial_scan' with 'no_initial_scan'.
The default behavior is to perform no initial scans
when new targets are added.

Release note (enterprise change): Users can now
perform initial scans on newly added targets
by executing the following statement:

ALTER CHANGEFEED <job_id> ADD <targets> WITH initial_scan

The default behavior is to perform no initial scans
on newly added targets, but users can explicitly
request this by replacing the 'initial_scan' with
'no_initial_scan'.

Release justification: low danger enhancement to newly
implemented functionality.

sherman-grewal commented:
bors r+
